### ISSUE TITLE
Interactively-defined function loses indentation when saved or edited

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#5284]
I first defined a function interactively at the prompt, not entering any indentation manually (i.e. fish indentend my code each time I pressed `Enter`):

```fish
mqudsi@Blitzkrieg /m/d/r/fish> function replace_in
                                   # argv[1] is the regex pattern of files to match, fed to `fd`
                                   # argv[2] is the pattern of text to search for
                                   # argv[3] is the pattern of text it should be replaced with
                                   for f in (fd $argv[1] .)
                                       string replace -a -- $argv[2] $argv[3] < $f | rewrite $f
                                   end
                               end
```

The function definition shows the indentation:

```fish
mqudsi@Blitzkrieg /m/d/r/fish> type replace_in
replace_in is a function with definition
function replace_in

    # argv[2] is the pattern of text to search for
    # argv[3] is the pattern of text it should be replaced with
    for f in (fd $argv[1] .)
        string replace -a -- $argv[2] $argv[3] <$f | rewrite $f
    end
end
```

`funced` without `$EDITOR` set shows the following:

```fish
replace_in> function replace_in
                
                # argv[2] is the pattern of text to search for
                # argv[3] is the pattern of text it should be replaced with
                for f in (fd $argv[1] .)
                    string replace -a -- $argv[2] $argv[3] <$f | rewrite $f
                end
            end
```

But this is what happens after I `set -Ux EDITOR vim`, then execute `funced replace_in` again:

![image](https://user-images.githubusercontent.com/606923/47609426-d77f4780-da03-11e8-8b67-752b41b2effb.png)

**Comment by ridiculousfish** [comment]
This is pretty weird. I wasn't able to reproduce it.

Might be due to vim config. What happens if you `set -x EDITOR cat`? Does `cat` show the function as indented or not?

**Comment by faho** [comment]
I can reproduce as well, but it doesn't have anything to do with `funced` - `functions replace_in` already exhibits the issue.

The reason seems pretty simple, actually. When you interactively define a function, fish lies about the indentation.

When you type

```fish
$ function something
       echo a
```

that `echo a` isn't _actually_ indented - when you go to before the `e` and do a single backward-char you end up in the line before.

We probably don't want to lose that latter property - you don't want to mess with indentation in the commandline buffer - so we could either add the apparent indentation later.

Or... and this is much simpler: We could just let `funced` (and also `funcsave`) go through `fish_indent`. Which "forces" people to use its indentation, but that's basically a feature.

@ridiculousfish: Did you define the function by typing it into the commandline?

**Comment by ridiculousfish** [comment]
Yes after typing in the command line I can reproduce. Now I get it, when defining a function on the command line, fish inserts apparent indentation.

`fish_indent` is heavyweight, a more lightweight approach is to use `parse_util_compute_indents` which is what the command line editor uses.

**Comment by lukaszmoroz** [comment]
The reverse problem happens when you feed the output of `fish_indent` to `commandline` - the indentation is doubled:

```fish
function __format_commandline
    commandline (commandline | fish_indent)
    # You actually need to do:
    # commandline (commandline | fish_indent | string trim)
end

bind \cO __format_commandline
```

**Comment by max-sixty** [comment]
@lukaszmoroz yes I found that too. 

The issue is arguably with how fish displays the command — it adds visual indentation to lines in the CLI. So running `fish_indent` on them applies actual whitespace, making it double-indented.

Ideally it would be great to have a flag that disables this additional visual indentation, so `commandline -r $"(commandline | fish_indent)"` worked without this double-indentation.

(Also note that `$(...)` is needed, to handle commands with line separators, IIUC)

**Comment by krobelus** [comment]
this is fixed by 611a0572b~3..611a0572b
To format the commandline one can use ` | fish_indent | fish_indent --only-unindent`.
Perhaps this is the wrong interface and we should add it to `commandline` instead but it's simple and testable
